File: cpp/Linked_List/0024_Swap_Nodes_in_Pairs.cpp
using namespace std;

struct ListNode {
  int val;
  ListNode *next;

  ListNode() : val(0), next(nullptr) {}

  ListNode(int x) : val(x), next(nullptr) {}

  ListNode(int x, ListNode *next) : val(x), next(next) {}
};
//leetcode submit region begin(Prohibit modification and deletion)
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
  /*
   * 1st solution recursive
   * */
  ListNode *swapPairs(ListNode *head) {
    if (!head || !head->next) return head;
    ListNode *temp = head->next;
    head->next = swapPairs(temp->next);
    temp->next = head;
    return temp;
  }

  /*
   * 2nd solution iterative
   */
  ListNode *swapPairs2(ListNode *head) {
    ListNode *dummyHead = new ListNode(0);
    dummyHead->next = head;
    ListNode *cur = dummyHead;

    while (cur->next && cur->next->next) {
      ListNode *temp1 = cur->next;
      ListNode *temp2 = cur->next->next->next;
      cur->next = cur->next->next;
      cur->next->next = temp1;
      cur->next->next->next = temp2;

      cur = cur->next->next;
    }

    return dummyHead->next;
  }
};
//leetcode submit region end(Prohibit modification and deletion)

File: python3/0024-swap-nodes-in-pairs.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def swapPairs(self, head: [ListNode]) -> [ListNode]:
        dummy_head = ListNode(0, head)
        pre = dummy_head

        while pre.next and pre.next.next:
            cur = pre.next
            post = pre.next.next

            #   0  ->  1  ->  2  ->  3
            #  pre    cur    post
            cur.next = post.next  # 1 -> 3
            post.next = cur  # 2 -> 1
            pre.next = post  # 0 -> 2

            pre = pre.next.next
        return dummy_head.next


def to_linked_list(iterable):
    head = None
    for val in reversed(iterable):
        head = ListNode(val, head)
    return head


def to_native_list(head):
    lst = []
    while head:
        lst.append(head.val)
        head = head.next
    return lst


head = [1, 2, 3, 4]
l1 = to_linked_list(head)
result = Solution().swapPairs(l1)
print(to_native_list(result))  # [2, 1, 4, 3]

head = []
l2 = to_linked_list(head)
result = Solution().swapPairs(l2)
print(to_native_list(result))  # []

head = [1]
l3 = to_linked_list(head)
result = Solution().swapPairs(l3)
print(to_native_list(result))  #[1]
